dbms/src/WindowFunctions/IWindowFunction.h

```diff
@@ -45,8 +45,25 @@ class IWindowFunction
 
 protected:
     DataTypes argument_types;
+
+private:
+    // Return true when the frame is valid
+    //
+    // Check if the frame is invalid, or we will insert null value into result column
+    // Frame is invalid when:
+    //   1. partition_end <= frame_start
+    //   2. frame_end < partition_start
+    virtual bool checkFrameValidAndHandle(WindowTransformAction &, size_t) { return true; }
 };
 
 using WindowFunctionPtr = std::shared_ptr<IWindowFunction>;
 
+// Runtime data for computing one window function.
+struct WindowFunctionWorkspace
+{
+    // TODO add aggregation function
+    WindowFunctionPtr window_function = nullptr;
+
+    ColumnNumbers arguments;
+};
```

Comment:
why not moving to `WindowUtils.h`?

Reply:
> why not moving to `WindowUtils.h`?

It relies on `WindowFunctionPtr` and the circle include will happen if we move this to `WindowUtils.h`

Reply:
ok